Synthetic _source: support match_only_text

nik9000 opened the pull request:
This adds support for synthetic `_source` to the `match_only_text` field
type. When synthetic `_source` is enabled `match_only_text` fields
create a hidden stored field to contain their text. This should have
similar or better search performance for this specific field type,
though it will have slightly worse indexing performance because
synthetic `_source` is still writing `_recovery_source`, which means
we're writing the bits for this field twice.

nik9000 committed:
Synthetic _source: support match_only_text

This adds support for synthetic `_source` to the `match_only_text` field
type. When synthetic `_source` is enabled `match_only_text` fields
create a hidden stored field to contain their text. This should have
similar or better search performance for this specific field type,
though it will have slightly worse indexing performance because
synthetic `_source` is still writing `_recovery_source`, which means
we're writing the bits for this field twice.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 committed:
Update docs/changelog/89516.yaml

nik9000 committed:
Words

nik9000 committed:
Fixup

nik9000 commented on modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java:
```diff
@@ -168,6 +185,35 @@ private Function<LeafReaderContext, CheckedIntFunction<List<Object>, IOException
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                };
```
There's a stored field lookup thing in `searchExecutionContext.lookup()` but it can't be convinced to load the hidden stored field. If we feel strongly about it I can try and integrate into it, but I'm not super sure how at the moment.

romseygeek replied:
Yeah that's strictly for scripts and is integrated poorly with other stored field lookup stuff, so I don't think it's worth trying to re-use it for the moment.  I *do* think that the document lookup API I'm playing with at the moment will improve this though.

nik9000 replied:
searchExecutionContext.lookup()

nik9000 replied:
It has a lovely caching mechanism that I think could be quite nice. If multiple queries need to recheck the source it'll load once while this won't.

romseygeek replied:
So, somewhat annoyingly, I don't think we will be able to re-use a stored field loader from the SearchLookup here because the underlying API expects a LeafReaderContext, not a LeafSearchLookup.  But you should be able to use a LeafStoredFieldLoader rather than a FieldVisitor here which will at least be more readable.

nik9000 committed:
Fixup

romseygeek commented on modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java:
```diff
@@ -326,6 +371,10 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
         context.addToFieldNames(fieldType().name());
+
+        if (context.isSyntheticSource()) {
+            context.doc().add(new Field(fieldType().originalFieldName(), value, ORIGINAL_FIELD_TYPE));
```
Use `StoredField` rather than creating a new field type

romseygeek commented on modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java:
```diff
@@ -279,6 +321,9 @@ public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext
             throw new IllegalArgumentException(CONTENT_TYPE + " fields do not support sorting and aggregations");
         }
 
+        private String originalFieldName() {
```
Maybe call this `storedFieldName` to make it a bit clearer what it's used for?

nik9000 replied:
:+1:

nik9000 replied:
:+1:

romseygeek commented on modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapperTests.java:
```diff
@@ -65,6 +91,13 @@ protected void registerParameters(ParameterChecker checker) throws IOException {
+        }
+        return fieldMapping(b -> b.field("type", "match_only_text"));
+    }
+
```
I think I'd be happier with coverage here if we explicitly run both tests for the synthetic and non-synthetic case, this feels at the moment like we're only testing 50% of the functionality on each run and I think that will come back to bite us.

nik9000 replied:
It has a lovely caching behavior that I kind of miss here. Like, if two queries need this then it'll cache.

romseygeek reviewed (commented):
This looks great! I think it may be worth waiting for a few days to see if I can get document loaders working though as it will tidy up the impl a fair amount.

nik9000 commented:
> This looks great! I think it may be worth waiting for a few days to see if I can get document loaders working though as it will tidy up the impl a fair amount.

:+1:

nik9000 committed:
More tests for enrich processor

Adds more tests for the enrich processor around different index types.
Right now they all work fine (yay!) but this feels like a good amount of
paranoia.

nik9000 committed:
Merge branch 'main' into synthetic_source_match_only_text

nik9000 committed:
Clean

nik9000 committed:
Merge branch 'main' into synthetic_source_match_only_text

nik9000 commented:
@romseygeek this should be ready for you too!

nik9000 committed:
Merge branch 'main' into synthetic_source_match_only_text

nik9000 committed:
Update

romseygeek reviewed (approved):
LGTM

nik9000 commented:
I almost clicked the merge button! It wouldn't have compiled! sneaky sneaky.

nik9000 committed:
Merge branch 'main' into synthetic_source_match_only_text

nik9000 committed:
Compile after merge

nik9000 committed:
Fixup
